File: src/endstone_endweave/protocol/v944_to_v924/handlers/block_pos.py
"""Packet handlers for BlockPos changes (v944 -> v924)."""

from endstone_endweave.codec import (
    BLOCK_POS,
    BOOL,
    BYTE,
    FLOAT_LE,
    INT_LE,
    ITEM_INSTANCE,
    NETWORK_BLOCK_POS,
    STRING,
    UVAR_INT,
    UVAR_INT64,
    VAR_INT,
    VAR_INT64,
    PacketWrapper,
)

_NOTE_BLOCK_EVENT = 0
_TRUMPET_INSERTION_POINT = 16
_TRUMPET_ID_SHIFT = 4


def _block_to_net(wrapper: PacketWrapper) -> None:
    wrapper.write(NETWORK_BLOCK_POS, wrapper.read(BLOCK_POS))


def _net_to_block(wrapper: PacketWrapper) -> None:
    wrapper.write(BLOCK_POS, wrapper.read(NETWORK_BLOCK_POS))


def _passthrough_inventory_action(wrapper: PacketWrapper) -> None:
    source_type = wrapper.passthrough(UVAR_INT)
    if source_type in (0, 99999):
        wrapper.passthrough(VAR_INT)
    elif source_type == 2:
        wrapper.passthrough(UVAR_INT)
    wrapper.passthrough(UVAR_INT)
    wrapper.passthrough(ITEM_INSTANCE)
    wrapper.passthrough(ITEM_INSTANCE)


def _passthrough_structure_settings_up(wrapper: PacketWrapper) -> None:
    wrapper.passthrough(STRING)
    wrapper.passthrough(BOOL)
    wrapper.passthrough(BOOL)
    wrapper.passthrough(BOOL)
    _net_to_block(wrapper)
    _net_to_block(wrapper)
    wrapper.passthrough(VAR_INT64)
    wrapper.passthrough(BYTE)
    wrapper.passthrough(BYTE)
    wrapper.passthrough(BYTE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(INT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)


def rewrite_first_block_to_net(wrapper: PacketWrapper) -> None:
    """Rewrite first-field BlockPos -> NetworkBlockPos.

    Args:
        wrapper: Packet wrapper positioned at the first field.
    """
    _block_to_net(wrapper)


def rewrite_tile_event(wrapper: PacketWrapper) -> None:
    """BlockEventPacket (26): convert BlockPos -> NetworkBlockPos, remap NoteBlockInstrument.

    Args:
        wrapper: Packet wrapper for TileEvent.
    """
    _block_to_net(wrapper)
    event_type = wrapper.passthrough(VAR_INT)
    event_data = wrapper.read(VAR_INT)
    if event_type == _NOTE_BLOCK_EVENT and event_data >= _TRUMPET_INSERTION_POINT + _TRUMPET_ID_SHIFT:
        event_data -= _TRUMPET_ID_SHIFT
    wrapper.write(VAR_INT, event_data)


def rewrite_set_spawn_position(wrapper: PacketWrapper) -> None:
    """SetSpawnPosition (43): convert BlockPos -> NetworkBlockPos.

    Args:
        wrapper: Packet wrapper for SetSpawnPosition.
    """
    wrapper.passthrough(VAR_INT)
    _block_to_net(wrapper)
    wrapper.passthrough(VAR_INT)
    _block_to_net(wrapper)


def rewrite_add_volume_entity(wrapper: PacketWrapper) -> None:
    """AddVolumeEntity (166): convert bounds BlockPos -> NetworkBlockPos.

    Args:
        wrapper: Packet wrapper for AddVolumeEntity.
    """
    wrapper.passthrough(UVAR_INT)
    _block_to_net(wrapper)
    _block_to_net(wrapper)


def rewrite_update_sub_chunk_blocks(wrapper: PacketWrapper) -> None:
    """UpdateSubChunkBlocks (172): convert all BlockPos fields.

    Args:
        wrapper: Packet wrapper for UpdateSubChunkBlocks.
    """
    _block_to_net(wrapper)
    blocks_count = wrapper.passthrough(UVAR_INT)
    for _ in range(blocks_count):
        _block_to_net(wrapper)
        wrapper.passthrough(UVAR_INT)
        wrapper.passthrough(UVAR_INT)
        wrapper.passthrough(UVAR_INT64)
        wrapper.passthrough(UVAR_INT)

    extra_count = wrapper.passthrough(UVAR_INT)
    for _ in range(extra_count):
        _block_to_net(wrapper)
        wrapper.passthrough(UVAR_INT)
        wrapper.passthrough(UVAR_INT)
        wrapper.passthrough(UVAR_INT64)
        wrapper.passthrough(UVAR_INT)


def rewrite_play_sound(wrapper: PacketWrapper) -> None:
    """PlaySound (86): convert BlockPos -> NetworkBlockPos.

    Args:
        wrapper: Packet wrapper for PlaySound.
    """
    wrapper.passthrough(STRING)
    _block_to_net(wrapper)


def rewrite_map_data(wrapper: PacketWrapper) -> None:
    """ClientboundMapItemData (67): convert tracked block object positions.

    Args:
        wrapper: Packet wrapper for ClientboundMapItemData.
    """
    wrapper.passthrough(VAR_INT64)
    types = wrapper.passthrough(UVAR_INT)
    wrapper.passthrough(BYTE)
    wrapper.passthrough(BOOL)
    wrapper.passthrough(NETWORK_BLOCK_POS)

    type_texture_update = 0x02
    type_decoration_update = 0x04
    type_creation = 0x08

    if types & type_creation:
        count = wrapper.passthrough(UVAR_INT)
        for _ in range(count):
            wrapper.passthrough(VAR_INT64)

    if types & (type_creation | type_decoration_update | type_texture_update):
        wrapper.passthrough(BYTE)

    if types & type_decoration_update:
        object_count = wrapper.passthrough(UVAR_INT)
        for _ in range(object_count):
            object_type = wrapper.passthrough(INT_LE)
            if object_type == 0:
                wrapper.passthrough(VAR_INT64)
            elif object_type == 1:
                _block_to_net(wrapper)


def rewrite_update_client_input_locks(wrapper: PacketWrapper) -> None:
    """UpdateClientInputLocks (196): append removed Server Pos.

    Args:
        wrapper: Packet wrapper for UpdateClientInputLocks.
    """
    wrapper.passthrough(UVAR_INT)
    wrapper.write(FLOAT_LE, 0.0)
    wrapper.write(FLOAT_LE, 0.0)
    wrapper.write(FLOAT_LE, 0.0)


def rewrite_inventory_transaction(wrapper: PacketWrapper) -> None:
    """Rewrite InventoryTransaction: convert NetworkBlockPos -> BlockPos in UseItem data.

    Args:
        wrapper: Packet wrapper for InventoryTransaction.
    """
    legacy_request_id = wrapper.passthrough(VAR_INT)
    if legacy_request_id != 0:
        slot_count = wrapper.passthrough(UVAR_INT)
        for _ in range(slot_count):
            wrapper.passthrough(BYTE)
            slots_len = wrapper.passthrough(UVAR_INT)
            for _ in range(slots_len):
                wrapper.passthrough(BYTE)

    transaction_type = wrapper.passthrough(UVAR_INT)
    action_count = wrapper.passthrough(UVAR_INT)
    for _ in range(action_count):
        _passthrough_inventory_action(wrapper)

    if transaction_type != 2:
        return

    wrapper.passthrough(UVAR_INT)
    wrapper.passthrough(UVAR_INT)
    _net_to_block(wrapper)
    wrapper.passthrough(VAR_INT)
    wrapper.passthrough(VAR_INT)
    wrapper.passthrough(ITEM_INSTANCE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(FLOAT_LE)
    wrapper.passthrough(UVAR_INT)
    wrapper.passthrough(UVAR_INT)
    wrapper.write(BYTE, 0)


def rewrite_player_action(wrapper: PacketWrapper) -> None:
    """PlayerAction (36): convert NetworkBlockPos -> BlockPos.

    Args:
        wrapper: Packet wrapper for PlayerAction.
    """
    wrapper.passthrough(UVAR_INT64)
    wrapper.passthrough(VAR_INT)
    _net_to_block(wrapper)
    _net_to_block(wrapper)


def rewrite_container_open(wrapper: PacketWrapper) -> None:
    """ContainerOpen (46): convert BlockPos -> NetworkBlockPos.

    Args:
        wrapper: Packet wrapper for ContainerOpen.
    """
    wrapper.passthrough(BYTE)
    wrapper.passthrough(BYTE)
    _block_to_net(wrapper)


def rewrite_structure_block_update(wrapper: PacketWrapper) -> None:
    """StructureBlockUpdate (90): convert NetworkBlockPos -> BlockPos in StructureSettings.

    Args:
        wrapper: Packet wrapper for StructureBlockUpdate.
    """
    _net_to_block(wrapper)
    wrapper.passthrough(STRING)
    wrapper.passthrough(STRING)
    wrapper.passthrough(BOOL)
    wrapper.passthrough(BOOL)
    wrapper.passthrough(VAR_INT)
    _passthrough_structure_settings_up(wrapper)
    wrapper.passthrough(VAR_INT)


def rewrite_command_block_update(wrapper: PacketWrapper) -> None:
    """CommandBlockUpdate (78): convert NetworkBlockPos -> BlockPos.

    Args:
        wrapper: Packet wrapper for CommandBlockUpdate.
    """
    is_block = wrapper.passthrough(BOOL)
    if is_block:
        _net_to_block(wrapper)


def rewrite_structure_template_data_request(wrapper: PacketWrapper) -> None:
    """StructureTemplateDataRequest (132): convert NetworkBlockPos -> BlockPos.

    Args:
        wrapper: Packet wrapper for StructureTemplateDataRequest.
    """
    wrapper.passthrough(STRING)
    _net_to_block(wrapper)
    _passthrough_structure_settings_up(wrapper)


def rewrite_anvil_damage(wrapper: PacketWrapper) -> None:
    """AnvilDamage (141): convert NetworkBlockPos -> BlockPos.

    Args:
        wrapper: Packet wrapper for AnvilDamage.
    """
    wrapper.passthrough(BYTE)
    _net_to_block(wrapper)

File: src/protocol/v924_to_v944/block_pos.cpp
/// v924 -> v944 block position handlers.
///
/// Clientbound: NetworkBlockPos -> BlockPos
/// Serverbound: BlockPos -> NetworkBlockPos

#include "endweave/protocol/v924_to_v944/handlers.h"

#include "endweave/codec/types/enums.h"
#include "endweave/codec/types/inventory.h"
#include "endweave/codec/types/item.h"
#include "endweave/codec/types/primitives.h"
#include "endweave/codec/types/structure.h"

namespace endweave::v924_to_v944 {

namespace {
constexpr int NOTE_BLOCK_EVENT = 0;
constexpr int TRUMPET_INSERTION_POINT = 16;
constexpr int TRUMPET_ID_SHIFT = 4;
} // namespace

// -- Clientbound: NetworkBlockPos -> BlockPos --

void rewrite_first_net_block_to_block(PacketWrapper& wrapper) {
    (void)wrapper.map<network_block_pos, block_pos>();
}

void rewrite_lectern_update_clientbound(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<byte_t>();  // New page to show
    (void)wrapper.passthrough<byte_t>();  // Total Pages
    (void)wrapper.map<network_block_pos, block_pos>();
}

void rewrite_lectern_update_serverbound(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<byte_t>();
    (void)wrapper.passthrough<byte_t>();
    (void)wrapper.map<block_pos, network_block_pos>();
}

void rewrite_tile_event(PacketWrapper& wrapper) {
    (void)wrapper.map<network_block_pos, block_pos>();  // Block Position
    auto event_type = wrapper.passthrough<var_int>();    // Event Type
    auto event_data = wrapper.read<var_int>();           // Event Value
    if (event_type && event_data) {
        auto val = *event_data;
        if (*event_type == NOTE_BLOCK_EVENT && val >= TRUMPET_INSERTION_POINT)
            val += TRUMPET_ID_SHIFT;
        wrapper.write<var_int>(val);
    }
}

void rewrite_set_spawn_position(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<var_int>();                // Spawn Position Type
    (void)wrapper.map<network_block_pos, block_pos>();   // Block Position
    (void)wrapper.passthrough<var_int>();                // Dimension type
    (void)wrapper.map<network_block_pos, block_pos>();   // Spawn Block Pos
}

void rewrite_add_volume_entity(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<uvar_int>();               // Entity Network Id
    (void)wrapper.map<network_block_pos, block_pos>();   // Min Bounds
    (void)wrapper.map<network_block_pos, block_pos>();   // Max Bounds
}

void rewrite_update_sub_chunk_blocks(PacketWrapper& wrapper) {
    (void)wrapper.map<network_block_pos, block_pos>();   // Sub Chunk Block Position

    // Blocks Changed - Standards
    auto blocks_count = wrapper.passthrough<uvar_int>();
    if (blocks_count) {
        for (std::uint32_t i = 0; i < *blocks_count; ++i) {
            (void)wrapper.map<network_block_pos, block_pos>();
            (void)wrapper.passthrough<uvar_int>();   // Runtime Id
            (void)wrapper.passthrough<uvar_int>();   // Update Flags
            (void)wrapper.passthrough<uvar_int64>(); // Sync Message - Entity Unique ID
            (void)wrapper.passthrough<uvar_int>();   // Sync Message - Message
        }
    }

    // Blocks Changed - Extras
    auto extra_count = wrapper.passthrough<uvar_int>();
    if (extra_count) {
        for (std::uint32_t i = 0; i < *extra_count; ++i) {
            (void)wrapper.map<network_block_pos, block_pos>();
            (void)wrapper.passthrough<uvar_int>();
            (void)wrapper.passthrough<uvar_int>();
            (void)wrapper.passthrough<uvar_int64>();
            (void)wrapper.passthrough<uvar_int>();
        }
    }
}

void rewrite_play_sound(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<string>();                 // Name
    (void)wrapper.map<network_block_pos, block_pos>();   // Position
}

void rewrite_map_data(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<var_int64>();   // Map ID
    auto types = wrapper.passthrough<uvar_int>();  // Type Flags
    (void)wrapper.passthrough<byte_t>();      // Dimension
    (void)wrapper.passthrough<bool_t>();      // Is Locked Map?
    (void)wrapper.passthrough<block_pos>();   // Map Origin

    if (!types) return;

    if (*types & static_cast<std::uint32_t>(ClientboundMapItemDataType::Creation)) {
        // Map ID List: uvarint count + varint64 entries
        auto count = wrapper.passthrough<uvar_int>();
        if (count) {
            for (std::uint32_t i = 0; i < *count; ++i)
                (void)wrapper.passthrough<var_int64>();
        }
    }

    if (*types & (static_cast<std::uint32_t>(ClientboundMapItemDataType::Creation) |
                  static_cast<std::uint32_t>(ClientboundMapItemDataType::DecorationUpdate) |
                  static_cast<std::uint32_t>(ClientboundMapItemDataType::TextureUpdate))) {
        (void)wrapper.passthrough<byte_t>(); // Scale
    }

    if (*types & static_cast<std::uint32_t>(ClientboundMapItemDataType::DecorationUpdate)) {
        auto obj_count = wrapper.passthrough<uvar_int>();
        if (obj_count) {
            for (std::uint32_t i = 0; i < *obj_count; ++i) {
                auto obj_type = wrapper.passthrough<int_le>();
                if (obj_type && *obj_type == static_cast<std::int32_t>(MapItemTrackedActorType::Entity)) {
                    (void)wrapper.passthrough<var_int64>();
                } else if (obj_type && *obj_type == static_cast<std::int32_t>(MapItemTrackedActorType::BlockEntity)) {
                    (void)wrapper.map<network_block_pos, block_pos>();
                }
            }
        }
    }
}

void rewrite_update_client_input_locks(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<uvar_int>(); // Input Lock ComponentData
    (void)wrapper.read<float_le>();        // discard Server Pos.X
    (void)wrapper.read<float_le>();        // discard Server Pos.Y
    (void)wrapper.read<float_le>();        // discard Server Pos.Z
}

// -- Serverbound: BlockPos -> NetworkBlockPos --

void rewrite_first_block_to_net_block(PacketWrapper& wrapper) {
    (void)wrapper.map<block_pos, network_block_pos>();
}

void rewrite_inventory_transaction(PacketWrapper& wrapper) {
    auto legacy_request_id = wrapper.passthrough<var_int>();
    if (legacy_request_id && *legacy_request_id != 0) {
        auto slot_count = wrapper.passthrough<uvar_int>();
        if (slot_count) {
            for (std::uint32_t i = 0; i < *slot_count; ++i) {
                (void)wrapper.passthrough<byte_t>();   // Container Enum
                auto slots_len = wrapper.passthrough<uvar_int>();
                if (slots_len) {
                    for (std::uint32_t j = 0; j < *slots_len; ++j)
                        (void)wrapper.passthrough<byte_t>(); // Slot
                }
            }
        }
    }

    auto transaction_type = wrapper.passthrough<uvar_int>();
    auto action_count = wrapper.passthrough<uvar_int>();
    if (action_count) {
        for (std::uint32_t i = 0; i < *action_count; ++i)
            (void)wrapper.passthrough<inventory_action>();
    }

    if (!transaction_type ||
        *transaction_type != static_cast<std::uint32_t>(ComplexInventoryTransactionType::ItemUseTransaction))
        return;

    // UseItemTransactionData
    (void)wrapper.passthrough<uvar_int>();                // ActionType
    (void)wrapper.passthrough<uvar_int>();                // TriggerType
    (void)wrapper.map<block_pos, network_block_pos>();    // BlockPosition
    (void)wrapper.passthrough<var_int>();                 // BlockFace
    (void)wrapper.passthrough<var_int>();                 // HotBarSlot
    (void)wrapper.passthrough<item_instance>();           // HeldItem
    (void)wrapper.passthrough<vec3>();                    // Position
    (void)wrapper.passthrough<vec3>();                    // ClickedPosition
    (void)wrapper.passthrough<uvar_int>();                // BlockRuntimeID
    (void)wrapper.passthrough<uvar_int>();                // ClientPrediction
    (void)wrapper.read<byte_t>();                         // ClientCooldownState (strip)
}

void rewrite_player_action(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<uvar_int64>(); // Player Runtime ID
    (void)wrapper.passthrough<var_int>();     // Action
    (void)wrapper.map<block_pos, network_block_pos>(); // Block Position
    (void)wrapper.map<block_pos, network_block_pos>(); // Result Pos
}

void rewrite_container_open(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<byte_t>(); // Container Id
    (void)wrapper.passthrough<byte_t>(); // Container Type
    (void)wrapper.map<network_block_pos, block_pos>(); // Position
}

void rewrite_structure_block_update(PacketWrapper& wrapper) {
    (void)wrapper.map<block_pos, network_block_pos>(); // Block Position
    // StructureEditorData
    (void)wrapper.passthrough<string>();  // Name
    (void)wrapper.passthrough<string>();  // DataField
    (void)wrapper.passthrough<bool_t>();  // IncludePlayers
    (void)wrapper.passthrough<bool_t>();  // ShowBoundingBox
    (void)wrapper.passthrough<var_int>(); // StructureBlockType
    (void)wrapper.map<structure_settings_v944, structure_settings_v924>();
}

void rewrite_command_block_update(PacketWrapper& wrapper) {
    auto is_block = wrapper.passthrough<bool_t>();
    if (is_block && *is_block) {
        (void)wrapper.map<block_pos, network_block_pos>();
    }
}

void rewrite_structure_template_data_request(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<string>();  // Structure Name
    (void)wrapper.map<block_pos, network_block_pos>(); // Structure Position
    (void)wrapper.map<structure_settings_v944, structure_settings_v924>();
}

void rewrite_anvil_damage(PacketWrapper& wrapper) {
    (void)wrapper.passthrough<byte_t>(); // Damage Amount
    (void)wrapper.map<block_pos, network_block_pos>(); // Block Position
}

} // namespace endweave::v924_to_v944
